tracing-subscriber/src/fmt/fmt_layer.rs

```diff
@@ -0,0 +1,532 @@
+        LayerBuilder {
+            fmt_fields: format::DefaultFields::default(),
+            fmt_event: format::Format::default(),
+            make_writer: io::stdout,
+            _inner: PhantomData,
+        }
+    }
+}
+
+impl<S, N, E, W> Layer<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    #[inline]
+    fn make_ctx<'a>(&'a self, ctx: Context<'a, S>) -> FmtContext<'a, S, N> {
+        FmtContext {
+            ctx,
+            fmt_fields: &self.fmt_fields,
+        }
+    }
+}
+
+/// A formatted representation of a span's fields stored in its [extensions].
+///
+/// Because `FormattedFields` is generic over the type of the formatter
+/// that produced it, multiple versions of a span's formatted fields can be
+/// stored in the [`Extensions`][extensions] type-map. This means that when
+/// multiple formatters are in use, each can store its own formatted
+/// representation without conflicting.
+///
+/// [extensions]: ../registry/extensions/index.html
+#[derive(Default)]
+pub struct FormattedFields<E> {
+    _format_event: PhantomData<fn(E)>,
+    /// The formatted fields of a span.
+    fields: String,
+}
+
+impl<E> fmt::Debug for FormattedFields<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FormattedFields")
+            .field("fields", &self.fields)
+            .finish()
+    }
+}
+
+impl<E> fmt::Display for FormattedFields<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.fields)
+    }
+}
+
+impl<E> Deref for FormattedFields<E> {
+    type Target = String;
+    fn deref(&self) -> &Self::Target {
+        &self.fields
+    }
+}
+
+// === impl FmtLayer ===
+
+impl<S, N, E, W> layer::Layer<S> for Layer<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    fn new_span(&self, attrs: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+
+        let mut buf = String::new();
+        if self.fmt_fields.format_fields(&mut buf, attrs).is_ok() {
+            let fmt_fields = FormattedFields {
+                fields: buf,
+                _format_event: PhantomData::<fn(N)>,
+            };
+            extensions.insert(fmt_fields);
+        }
+    }
+
+    fn on_record(&self, id: &Id, values: &Record<'_>, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+
+        let mut buf = String::new();
+        if self.fmt_fields.format_fields(&mut buf, values).is_ok() {
+            let buf = match extensions.get_mut::<FormattedFields<Self>>() {
```

Comment:
this will _always_ return `None`, since you are parameterizing the `FormattedFields` with `Self` (a `FormatEvent` implementation), but whenever you insert into the extensions, you pass `N` (the `FormatFields` implementation) as the type parameter, here and here:
https://github.com/tokio-rs/tracing/blob/5cc2e38c2b4eab1d65cb2e0abb8453c9a9436c9e/tracing-subscriber/src/fmt/fmt_layer.rs#L335-L339

https://github.com/tokio-rs/tracing/blob/5cc2e38c2b4eab1d65cb2e0abb8453c9a9436c9e/tracing-subscriber/src/fmt/fmt_layer.rs#L353-L357

we should change this to also use `N`:
```suggestion
            let buf = match extensions.get_mut::<FormattedFields<N>>() {
```

Reply:
also...do you think there should maybe there should be a unit test or two for this? since it seems like there are a few bugs in here...

Reply:
I made the suggested fix, thanks.

How do you suggest testing this one? Any examples you can point me towards?

Reply:
we could have a test that creates a span with some fields, and then records some new values, and make assertions about what the formatted fields look like...i am not going to block the PR on a test for this...